File: src/2025-02-27-deepclaude37-rust/go/exercises/practice/variable-length-quantity/variable_length_quantity.go
package variablelengthquantity

import "errors"

func EncodeVarint(input []uint32) []byte {
	var encoded []byte
	for _, num := range input {
		if num == 0 {
			encoded = append(encoded, 0x00)
			continue
		}
		var chunks []byte
		for num > 0 {
			chunk := byte(num & 0x7F)
			chunks = append(chunks, chunk)
			num >>= 7
		}
		// Reverse chunks to get most significant part first
		for i, j := 0, len(chunks)-1; i < j; i, j = i+1, j-1 {
			chunks[i], chunks[j] = chunks[j], chunks[i]
		}
		// Set continuation bits on all but last chunk
		for i := 0; i < len(chunks)-1; i++ {
			chunks[i] |= 0x80
		}
		encoded = append(encoded, chunks...)
	}
	return encoded
}

func DecodeVarint(input []byte) ([]uint32, error) {
	var result []uint32
	var current uint32
	var hasPending bool
	for _, b := range input {
		current = (current << 7) | uint32(b&0x7F)
		if (b & 0x80) == 0 {
			result = append(result, current)
			current = 0
			hasPending = false
		} else {
			hasPending = true
		}
	}
	if hasPending {
		return nil, errors.New("incomplete sequence")
	}
	return result, nil
}

File: src/2025-03-16-08-42-59--a-deepclaude-full-python-0316-1/python/exercises/practice/variable-length-quantity/variable_length_quantity.py
def encode(numbers):
    """
    Encode a list of integers using Variable Length Quantity encoding.
    
    Each byte uses 7 bits for data and 1 bit (MSB) to indicate if more bytes follow.
    """
    result = []
    for number in numbers:
        # Special case for zero
        if number == 0:
            result.append(0)
            continue
            
        # Break the number into 7-bit chunks
        chunks = []
        while number > 0:
            chunks.insert(0, number & 0x7F)
            number >>= 7
            
        # Set continuation bit (bit 7) on all but the last byte
        for i in range(len(chunks) - 1):
            chunks[i] |= 0x80
            
        result.extend(chunks)
    
    return result


def decode(bytes_):
    """
    Decode a Variable Length Quantity encoded sequence of bytes.
    
    Raises ValueError if the sequence is incomplete.
    """
    result = []
    current = 0
    
    for i, byte in enumerate(bytes_):
        # Add the 7 bits of data to our accumulator
        current = (current << 7) | (byte & 0x7F)
        
        # If MSB is clear, we've reached the end of a number
        if not (byte & 0x80):
            result.append(current)
            current = 0
        # Check if we're at the end of the bytes with an incomplete sequence
        elif i == len(bytes_) - 1:
            raise ValueError("incomplete sequence")
            
    # If we've processed all bytes but still have bits in current,
    # the sequence was incomplete
    if current != 0:
        raise ValueError("incomplete sequence")
            
    return result
